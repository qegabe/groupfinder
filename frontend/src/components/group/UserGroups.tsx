import React, { useEffect, useState } from "react";
import { Box, Typography } from "@mui/material";
import GroupList from "./GroupList";
import GroupFinderApi from "../../api";
import LoadingSpinner from "../common/LoadingSpinner";
import { useAppSelector } from "../../hooks/redux";

function UserGroups() {
  const user = useAppSelector((s) => s.auth.user);
  const [groupsData, setGroupsData] = useState<ListGroup[]>([]);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    async function loadGroups() {
      setGroupsData(
        await GroupFinderApi.getGroups({ hasUsers: [user?.username] })
      );
      setLoading(false);
    }
    loadGroups();
  }, [user?.username]);

  return (
    <Box sx={{ my: 4, mx: 2 }}>
      <Typography variant="h4">Your Groups</Typography>
      {loading ? <LoadingSpinner /> : <GroupList groupsData={groupsData} />}
    </Box>
  );
}

export default UserGroups;
