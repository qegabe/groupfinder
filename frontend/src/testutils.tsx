import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

function customRender(
  ui: ReactElement,
  route?: string,
  preloadedState?: PreloadedState<RootState>,
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter initialEntries={route ? [route] : undefined}>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {children}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
