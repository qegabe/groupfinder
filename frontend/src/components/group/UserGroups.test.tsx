import React from "react";
import { render, waitFor, screen } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import dayjs from "dayjs";
import UserGroups from "./UserGroups";

const mockGetGroups = jest
  .spyOn(GroupFinderApi, "getGroups")
  .mockImplementation(async () => {
    return [
      {
        id: 1,
        title: "Test Group 1",
        startTime: dayjs("2023-06-23T16:00:00.000Z"),
        endTime: dayjs("2023-06-23T17:00:00.000Z"),
      },
      {
        id: 2,
        title: "Test Group 2",
        startTime: dayjs("2023-06-23T17:00:00.000Z"),
        endTime: dayjs("2023-06-23T18:00:00.000Z"),
      },
      {
        id: 3,
        title: "Test Group 3",
        startTime: dayjs("2023-06-23T18:00:00.000Z"),
        endTime: dayjs("2023-06-23T19:00:00.000Z"),
      },
    ];
  });

afterEach(() => {
  jest.clearAllMocks();
});

it("renders", async () => {
  const { container } = render(<UserGroups />, {
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(mockGetGroups).toHaveBeenCalledWith({ hasUsers: ["TestUser"] });
  expect(screen.getByText("Test Group 1")).toBeInTheDocument();
});

it("matches snapshot", async () => {
  const { container, asFragment } = render(<UserGroups />, {
    preloadedState: {
      auth: {
        token: null,
        user: { username: "TestUser", avatarUrl: null },
        error: null,
        loading: false,
      },
    },
  });
  await waitFor(() => {
    expect(container.querySelector(".LoadingSpinner")).not.toBeInTheDocument();
  });
  expect(asFragment()).toMatchSnapshot();
});
