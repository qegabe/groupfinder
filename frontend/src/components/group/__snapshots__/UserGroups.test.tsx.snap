// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-zitybv"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
    >
      Your Groups
    </h4>
    <div
      class="MuiStack-root css-1p5q5e5-MuiStack-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Test Group 1
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Starting 3 days ago
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            href="/groups/1"
            tabindex="0"
          >
            Details
          </a>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Test Group 2
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Starting 3 days ago
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            href="/groups/2"
            tabindex="0"
          >
            Details
          </a>
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 css-2ulfj5-MuiTypography-root"
          >
            Test Group 3
          </h6>
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Starting 3 days ago
          </p>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-1t6e9jv-MuiCardActions-root"
        >
          <a
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall css-1knaqv7-MuiButtonBase-root-MuiButton-root"
            href="/groups/3"
            tabindex="0"
          >
            Details
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
